pkg/security/security.go

```diff
@@ -167,21 +231,100 @@ func (p *provider) Handler(ctx context.Context) (Handler, error) {
 	}
 }
 
+// TLSServerConfigNoClientAuth returns a TLS server config which instruments
+// using the current signed server certificate. Does no client authentication.
+func (s *security) TLSServerConfigNoClientAuth() *tls.Config {
+	return tlsconfig.TLSServerConfig(s.source)
+}
+
+// GRPCDialOption returns a gRPC dial option which instruments client
+// authentication using the current signed client certificate.
+func (s *security) GRPCDialOption(appID spiffeid.ID) grpc.DialOption {
+	if !s.mtls {
+		return grpc.WithTransportCredentials(insecure.NewCredentials())
+	}
+	return grpc.WithTransportCredentials(
+		grpccredentials.MTLSClientCredentials(s.source, s.source, tlsconfig.AuthorizeID(appID)),
+	)
+}
+
 // GRPCServerOptionNoClientAuth returns a gRPC server option which instruments
 // authentication of clients using the current trust anchors. Doesn't require
 // clients to present a certificate.
 func (s *security) GRPCServerOptionNoClientAuth() grpc.ServerOption {
-	return grpc.Creds(
-		grpccredentials.TLSServerCredentials(s.source),
+	return grpc.Creds(grpccredentials.TLSServerCredentials(s.source))
+}
+
+// GRPCDialOptionUnknownTrustDomain returns a gRPC dial option which
+// instruments client authentication using the current signed client
+// certificate. Doesn't verify the servers trust domain, but does authorize the
+// SPIFFE ID path.
+// Used for clients which don't know the servers Trust Domain.
+func (s *security) GRPCDialOptionUnknownTrustDomain(ns, appID string) grpc.DialOption {
+	if !s.mtls {
+		return grpc.WithTransportCredentials(insecure.NewCredentials())
+	}
+
+	expID := fmt.Sprintf("/ns/%s/%s", ns, appID)
+	matcher := func(actual spiffeid.ID) error {
+		if actual.Path() != expID {
+			return fmt.Errorf("unexpected SPIFFE ID: %q", actual)
+		}
+		return nil
+	}
+
+	return grpc.WithTransportCredentials(
+		grpccredentials.MTLSClientCredentials(s.source, s.source, tlsconfig.AdaptMatcher(matcher)),
 	)
 }
 
+// WatchTrustAnchors watches for changes to the trust domains and returns the
+// PEM encoded trust domain roots.
+// Returns when the given context is canceled.
+func (s *security) WatchTrustAnchors(ctx context.Context, trustAnchors chan<- []byte) {
+	sub := make(chan struct{})
+	s.source.lock.Lock()
+	s.source.trustAnchorSubscribers = append(s.source.trustAnchorSubscribers, sub)
+	s.source.lock.Unlock()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-sub:
+			caBundle, err := s.CurrentTrustAnchors()
+			if err != nil {
+				log.Errorf("failed to marshal trust anchors: %s", err)
```

Comment:
failed to marshal?

Reply:
Yes:

```go
// CurrentTrustAnchors returns the current trust anchors for this Dapr
// installation.
func (s *security) CurrentTrustAnchors() ([]byte, error) {
	if s.source == nil {
		return nil, nil
	}
	return s.source.trustAnchors.Marshal()
}
```

Reply:
Ok. A nit would be to wrap the error in `CurrentTrustAnchors` since the name of the method doesn't imply a connection with marshalling